packages/pluggableWidgets/barcode-scanner-web/src/components/BarcodeScanner.tsx

```diff
@@ -0,0 +1,119 @@
+import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
+import classNames from "classnames";
+import { Alert } from "@mendix/piw-utils-internal";
+import { useCodeScanner, CodeScannerHookError } from "../hooks/useCodeScanner";
+import { browserSupportsCameraAccess, useMediaStream, MediaStreamHookError } from "../hooks/useMediaStream";
+
+import "../ui/BarcodeScanner.scss";
+
+export interface BarcodeScannerProps {
+    onClose?: () => void;
+    onDetect?: (data: string) => void;
+    showMask: boolean;
+}
+
+function getErrorMessage(errorEnum: MediaStreamHookError | CodeScannerHookError | null): string | null {
+    switch (errorEnum) {
+        case "ERROR_NOT_FOUND":
+            return "Error in barcode scanner: no camera media devices were found.";
+        case "ERROR_MEDIA_STREAM":
+            return "Error in barcode scanner: an unexpected error occurred while retrieving the camera media stream.";
+        case "ERROR_CODE_SCANNER":
+            return "Error in barcode scanner: an unexpected error occurred while detecting a barcode in the camera media stream.";
+        case "ERROR_NOT_ALLOWED":
+        default:
+            return null;
+    }
+}
+
+export function BarcodeScanner({ onClose, onDetect, showMask }: BarcodeScannerProps): ReactElement | null {
+    const [showScannerOverlay, setShowScannerOverlay] = useState<boolean>(true);
+    const [videoElement, setVideoElement] = useState<HTMLVideoElement | null>(null);
+    const hasDetectedOnce = useRef<boolean>(false);
+    const { streamObject, cleanupStreamObject, error: errorMediaStream } = useMediaStream();
+    const { codeResult, error: errorCodeScanner } = useCodeScanner(streamObject, videoElement);
+    const supportsCameraAccess = browserSupportsCameraAccess();
+
+    const updateVideoElement = useCallback(
+        (node: HTMLVideoElement | null) => {
+            setVideoElement(node);
+        },
+        [setVideoElement]
+    );
+
+    const play = useCallback((event: React.SyntheticEvent<HTMLVideoElement, Event>) => {
+        if (event.currentTarget.paused) {
+            event.currentTarget.play();
+        }
+    }, []);
+
+    function onCloseOverlay(): void {
+        setShowScannerOverlay(false);
+        cleanupStreamObject();
+    }
+
+    // If both the video element ref and the camera stream object are ready, display it through the `srcObject` prop.
+    useEffect(() => {
+        if (videoElement && streamObject) {
+            videoElement.srcObject = streamObject;
+        }
+    }, [streamObject, videoElement]);
+
+    // If we have an onDetect handler and a barcode has been scanned and it was the first detected code, trigger the onDetect handler.
+    useEffect(() => {
+        if (onDetect && codeResult && !hasDetectedOnce.current) {
```

Comment:
Isn't better to throttle the execution?

Reply:
I've changed it in e27c6b4 according to https://github.com/zxing-js/library/blob/master/src/browser/BrowserCodeReader.ts#L141